stdlib/public/Reflection/TypeRefBuilder.cpp

```diff
@@ -263,23 +309,32 @@ RemoteRef<FieldDescriptor> TypeRefBuilder::getFieldTypeInfo(const TypeRef *TR) {
   if (Found != FieldTypeInfoCache.end())
     return Found->second;
 
+  if (auto FD = getFieldDescriptorFromProvider(*MangledName))
+    return *FD;
+
   // Heuristic: find the outermost Module node available, and try to parse the
   // ReflectionInfos with a matching name first.
   auto ModuleName = FindOutermostModuleName(Node);
   // If we couldn't find a module name or the type is imported (__C module) we
   // don't any useful information on which image to look for the type.
   if (ModuleName && ModuleName != llvm::StringRef("__C"))
     for (size_t i = 0; i < ReflectionInfos.size(); ++i)
-      if (llvm::is_contained(ReflectionInfos[i].PotentialModuleNames,
-                             ModuleName))
-        if (auto FD = findFieldDescriptorAtIndex(i, *MangledName))
-          return *FD;
+      if ((!ExternalProvider ||
+           ExternalProvider->shouldParseReflectionInfoWithId(i)))
+        if (llvm::is_contained(ReflectionInfos[i].PotentialModuleNames,
```

Comment:
I think this could use another comment explaining the algorithm and the contract with ExternalProvider here. I'm not sure it's obvious to the reader that !shouldParseReflectionInfoWithId(i) means we already parsed it.